File: src/160.intersection-of-two-linked-lists/solution_two-pointers.cpp
/*
 * @lc app=leetcode.cn id=160 lang=cpp
 *
 * https://leetcode-cn.com/problems/intersection-of-two-linked-lists/
 */

/**
 * Definition for singly-linked list.
 */
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(nullptr) {}
};

// @lc code=start

class Solution {
 public:
  ListNode *getIntersectionNode(ListNode *l1, ListNode *l2) {
    if (!l1 || !l2) {
      return nullptr;
    }

    ListNode *p1 = l1;
    ListNode *p2 = l2;

    while (p1 != p2) {
      p1 = p1 ? p1->next : l2;
      p2 = p2 ? p2->next : l1;
    }

    return p1;
  }
};

// @lc code=end

File: src/61.rotate-list/solution_two-pointers.ts
/*
 * @lc app=leetcode.cn id=61 lang=typescript
 *
 * https://leetcode-cn.com/problems/rotate-list/
 */

/**
 * Definition for singly-linked list.
 */
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

// @lc code=start
// k >= 0, int
function rotateRight(head: ListNode | null, k: number): ListNode | null {
  if (!head) {
    return null;
  }

  const sentry = new ListNode(0, head);

  // resolved: original tail
  let p1 = sentry;
  // resolved: new tail
  let p2 = sentry;

  for (let i = 0; i < k; ++i) {
    p1 = p1.next as ListNode;

    // If p1 points to the original tail after this for-loop,
    // p2 will point to sentry node finally.
    // The sentry node cannot be the new tail node,
    // so we need to skip this case.
    if (!p1.next) {
      p1 = sentry;
    }
  }
  while (p1.next) {
    p1 = p1.next;
    p2 = p2.next as ListNode;
  }

  const answer = p2.next || head;

  p1.next = head;
  p2.next = null;

  return answer;
}
// @lc code=end

export default rotateRight;
